onnxscript/irbuilder.py

```diff
@@ -120,14 +125,48 @@ def __str__(self):
         return helper.printable_attribute(self.attr_proto)
 
     @property
-    def name(self):
+    def name(self) -> str:
         return self.attr_proto.name
 
     @property
-    def type(self):
+    def type(self) -> onnx.AttributeProto.AttributeType:
         return self.attr_proto.type
 
 
+@dataclasses.dataclass(frozen=True)
+class IRAttributeParameter:
+    """An attribute parameter (representing a formal parameter).
+
+    It may or may not carry a default value.
+
+    Attributes:
+        name: The name of the attribute.
+        type: The type of the attribute.
+        has_default: Whether the attribute has a default value.
+        attr_proto: The attribute proto.
+    """
+
+    name: str
+    type: onnx.AttributeProto.AttributeType
+    default_value: str | int | float | None = None
+
+    def __str__(self):
+        return helper.printable_attribute(self.attr_proto)
+
+    @property
+    def has_default(self):
+        return self.default_value is not None
+
+    @property
+    def attr_proto(self) -> onnx.AttributeProto:
+        if self.default_value is None:
```

Comment:
Is this method used in any context where the default_value is None? Wonder if we should throw an exception instead, since the returned value is an incomplete AttributeProto.

Reply:
OK!

Reply:
Done